Initial VMware Support

alekssaul opened the pull request:
# Initial VMware support for Tectonic 1.6.2 

## Scenarios tested
- Provisioning with self-hosted / experimental flags
- Provisioning with standalone etcd cluster.
- Scaling up the cluster by increasing variables such as master_count or worker_count
- Destroying a Virtual Machine by hand and recovering it via `terraform apply`
## Highlights
- End user brings in VMware CoreOS container image that's supported by the OS team. Currently Terraform does not have support for manipulating VMware virtual machine images/templates. It's left up to the operator to import this image (documented) and keep it up to date
- End user created DNS records required (documented). There is no universal DNS api for on-premise deployments.
- IP addresses are assigned statically using terraform map variables
- All machines are assumed to be deployed into the same Network subnet.

## Modules and Platform additions

Platforms follow similar syntax to existing supported platforms:
- tectonic.tf : contains use of bootkube and tectonic modules based mostly on AWS and Metal as examples.
- provider.tf : sets up vSphere provider configuration and creates a folder to deploy resources to. 
- remote.tf : scp assets generated similar to Baremetal flow. However copying of kubeconfig has been moved to individual node creation within the modules
- main.tf : uses node modules and etcd modules to create etcd and master-worker nodes. Master and worker generation are both handled by node module.

Modules/vmware contains two modules:
- etcd : Sets up an etcd cluster. Platform specific changes are:
  * Enable VMware Guest Tools, which provides hypervisor integrations. Guest tools are shipped with CoreOS Container Linux VMware image
  * Setup static networking
  * Setup etcd cluster by setting `--initial-cluster` flag to rendering of template_file. This produces static list of nodes based on end users input of etcd node count and fully qualified hostnames
- nodes: Sets up worker and master nodes. Platform Specific changes are:
  * Enable VMware Guest Tools, which provides hypervisor integrations. Guest tools are shipped with CoreOS Container Linux VMware image
  * Setup static networking
  * Uses `remote-exec` and `file` provisioners to copy kubeconfig similar to Baremetal setup

s-urbaniak commented on Documentation/variables/vmware.md:
```diff
@@ -0,0 +1,39 @@
+
+## Inputs
+
+| Name | Description | Default | Required |
```
We don't render the `Required` column anymore, as its semantics are dubious. Do you mind to regenerate this with a newer version of `terraform-docs` as outlined in https://github.com/coreos/tectonic-installer/blob/master/Makefile#L40-L45 ?

s-urbaniak commented on Documentation/variables/vmware.md:
```diff
@@ -0,0 +1,39 @@
+
+| Name | Description | Default | Required |
+|------|-------------|:-----:|:-----:|
+| tectonic_ssh_authorized_key | SSH public key to use as an authorized key. Example: `"ssh-rsa AAAB3N..."` | - | yes |
```
I suggest to call this `tectonic_vmware_ssh_authorized_key`

alekssaul replied:
I've used the variable without "vmware" prefix because of the usage in metal; https://github.com/coreos/tectonic-installer/blob/master/platforms/metal/variables.tf#L179-L187 . Happy to make the change if there are no plans to roll `tectonic_ssh_authorized_key` into global variables.

s-urbaniak commented on examples/terraform.tfvars.vmware:
```diff
@@ -0,0 +1,178 @@
+//  "1" = "etcd1"
+//  "2" = "etcd2"
+//  "3" = "etcd3"
+//  "4" = "etcd5"
```
I believe you added this manually. Since we are auto-generating this, I suggest to add a HEREDOC in the variable declaration below.

s-urbaniak commented on modules/vmware/etcd/ignition.tf:
```diff
@@ -0,0 +1,116 @@
+  count = "${length(var.external_endpoints) == 0 ? 1 : 0}"
+
+  name   = "locksmithd.service"
+  enable = true
```
We mask locksmith in favor of CLUO.

alekssaul replied:
Thanks @s-urbaniak , this is also enabled in AWS; https://github.com/coreos/tectonic-installer/blob/master/modules/aws/etcd/ignition.tf#L25-L37 . Also I believe CLUO is deployed to master& worker nodes, not etcd nodes?

s-urbaniak replied:
ah, you are right and I was blind ;-) Indeed, etcd nodes are not under control of CLUO.

s-urbaniak commented on modules/vmware/etcd/nodes.tf:
```diff
@@ -0,0 +1,27 @@
+resource "vsphere_virtual_machine" "etcd_node" {
+  count           = "${length(var.external_endpoints) == 0 ? var.instance_count : 0}"
```
Additionally if `experimental` is enabled, we deploy self-hosted etcd, hence etcd nodes don't need to be created in this case.

alekssaul replied:
`experimental` flag is handled via; https://github.com/alekssaul/tectonic-installer/blob/vmware-1.6/platforms/vmware/main.tf#L3 within the platform instead of the module. This I believe is similar to AWS; https://github.com/coreos/tectonic-installer/blob/master/modules/aws/etcd/nodes.tf#L26 and https://github.com/coreos/tectonic-installer/blob/master/platforms/aws/main.tf#L56

s-urbaniak replied:
correct, my blindness is quite shocking at this point ;-)

alekssaul replied:
you call it blindness, I call it Mondays :)

s-urbaniak commented on modules/vmware/etcd/nodes.tf:
```diff
@@ -0,0 +1,27 @@
+  disk {
+    datastore     = "${var.vm_disk_datastore}"  
+    template      = "${var.vm_disk_template_folder}/${var.vm_disk_template}"
+    type	        = "thin"
```
This is misaligned, I suggest to invoke `terraform fmt` on all files.

alekssaul commented:
Thanks @s-urbaniak for your review. I've made following changes based on your feedback.
- modify the Makefile to add vmware docs and examples
  * broke Makefile during the process and fixed (sorry about multiple commits)
- Added HEREDOC to variables file instead
- cleaned up indentation by running `terraform fmt`
- auto-generated docs and examples via `make docs ; make examples`
- change `tectonic_ssh_authorized_key` to `tectonic_vmware_ssh_authorized_key`

alekssaul commented:
+ Bootkube content_hash variable changed related to; https://github.com/coreos/tectonic-installer/pull/575

alekssaul commented:
Hi @s-urbaniak , 

Whenever you are available, I believe this is ready for review :)

s-urbaniak commented on platforms/vmware/variables.tf:
```diff
@@ -0,0 +1,212 @@
+  description = "Username to Use to connect to vCenter"
+}
+
+variable "tectonic_vmware_password" {
```
Is there any possibility to reference this and potentially the username outside of terraform? Currently we generally reference these sorts of credentials outside of terraform itself.

The discussion per se is in flux (see https://github.com/coreos/tectonic-installer/issues/710 is the umbrella issue) but as long as the discussion is going we would be more flexible referencing outside credentials first.

alekssaul commented:
Thanks @s-urbaniak 
- removed credentials from variables.tf
- Re-ran `make examples` and `make docs`
- Updated the documentation noting that end user will have to provide vmware credentials at runtime.

alekssaul commented:
@MattMencel was testing this in his environment and noticed that Master node provisioning was using CPU/memory size variables meant for Workers. https://github.com/coreos/tectonic-installer/pull/639/commits/b662c7140f8db0ec3b6f5354195b4fc4d6843063 fixes this.

alekssaul commented:
Hi @s-urbaniak, 

I removed the credentials from variables files. Users will be prompted during runtime (or they set `VSPHERE_USER` & `VSPHERE_PASSWORD` env variables).

Please let me know if there are any other concerns.

@alexsomesan

s-urbaniak commented:
@alekssaul from my side this is looking good! please re-execute `make docs examples` as these are out of sync causing the CI to fail.

alekssaul commented:
@s-urbaniak broke it first then fixed it :) thanks for help

alekssaul commented:
with merging of master into VMware branch the PR was broken

- https://github.com/coreos/tectonic-installer/pull/639/commits/1b18905b3d23ef20121092c3e615afa7ccc70863 adds changes required to support https://github.com/coreos/tectonic-installer/pull/767 
- https://github.com/coreos/tectonic-installer/pull/639/commits/c3e5ff305168e10bbfbaecb062e5707e08bd512c adds changes required for https://github.com/coreos/tectonic-installer/pull/829
- https://github.com/coreos/tectonic-installer/pull/639/commits/508a1db0fc6d93472ae09bc6ce00a6d066e30d93 add changes required for https://github.com/coreos/tectonic-installer/pull/805

alekssaul commented:
Master had 3 breaking changes ^^^ that needed to be applied to VMware platform.

s-urbaniak commented:
@alekssaul cool, thanks!

just a very last nit, CI tells me this failure:
```
misformatted files (run 'terraform fmt .' to fix): platforms/vmware/tectonic.tf
```

I guess you will have to invoke `terraform fmt` on the offending file.

s-urbaniak commented:
and once CI is green we shall merge this finally ;)

alekssaul commented:
@s-urbaniak all green

alexsomesan commented:
Way cool!

On Thu, Jun 1, 2017 at 6:30 AM, Sergiusz Urbaniak <notifications@github.com>
wrote:

> Merged #639 <https://github.com/coreos/tectonic-installer/pull/639>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/coreos/tectonic-installer/pull/639#event-1106039857>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABC-mx3PwrgWZuujaZrx3IKc9dTs_am-ks5r_rzzgaJpZM4NXS5M>
> .
>
